src/storage/sql_table.cpp

```diff
@@ -1,34 +1,324 @@
+        break;
+      case 2:
+        num_2_byte_attrs++;
+        break;
+      case 1:
+        num_1_byte_attrs++;
+        break;
+      case VARLEN_COLUMN:
+        num_varlen_byte_attrs++;
+      default:
+        break;
+    }
+  }
+
+  TERRIER_ASSERT(static_cast<uint16_t>(attr_sizes.size()) ==
+                     num_8_byte_attrs + num_4_byte_attrs + num_2_byte_attrs + num_1_byte_attrs + num_varlen_byte_attrs,
+                 "Number of attr_sizes does not match the sum of attr counts.");
+
+  // Initialize the offsets for each attr_size
+  auto offset_varlen_byte_attrs = static_cast<uint16_t>(NUM_RESERVED_COLUMNS);
+  auto offset_8_byte_attrs = static_cast<uint16_t>(offset_varlen_byte_attrs + num_varlen_byte_attrs);
+  auto offset_4_byte_attrs = static_cast<uint16_t>(offset_8_byte_attrs + (num_8_byte_attrs - NUM_RESERVED_COLUMNS));
+  auto offset_2_byte_attrs = static_cast<uint16_t>(offset_4_byte_attrs + num_4_byte_attrs);
+  auto offset_1_byte_attrs = static_cast<uint16_t>(offset_2_byte_attrs + num_2_byte_attrs);
+
+  ColumnMap col_map;
+  InverseColumnMap inv_col_map;
+
+  // Build the maps between Schema column OIDs and underlying column IDs
+  for (const auto &column : schema.GetColumns()) {
+    switch (column.GetAttrSize()) {
+      case 8:
+        inv_col_map[col_id_t(offset_8_byte_attrs)] = column.GetOid();
+        col_map[column.GetOid()] = col_id_t(offset_8_byte_attrs++);
+        break;
+      case 4:
+        inv_col_map[col_id_t(offset_4_byte_attrs)] = column.GetOid();
+        col_map[column.GetOid()] = col_id_t(offset_4_byte_attrs++);
+        break;
+      case 2:
+        inv_col_map[col_id_t(offset_2_byte_attrs)] = column.GetOid();
+        col_map[column.GetOid()] = col_id_t(offset_2_byte_attrs++);
+        break;
+      case 1:
+        inv_col_map[col_id_t(offset_1_byte_attrs)] = column.GetOid();
+        col_map[column.GetOid()] = col_id_t(offset_1_byte_attrs++);
+        break;
+      case VARLEN_COLUMN:
+        inv_col_map[col_id_t(offset_varlen_byte_attrs)] = column.GetOid();
+        col_map[column.GetOid()] = col_id_t(offset_varlen_byte_attrs++);
+        break;
+      default:
+        throw std::runtime_error("unexpected switch case value");
+    }
+  }
+
+  BlockLayout layout = storage::BlockLayout(attr_sizes);
+
+  auto dt = new DataTable(block_store_, layout, schema.GetVersion());
+  // clang's memory analysis has a false positive on this allocation.  The TERRIER_ASSERT on the second line of this
+  // function prevents the insert below from failing (can only fail when key is not unique).  The write-lock on the
+  // catalog prevents any other transaction from being in a race condition with this one.  The corresponding delete
+  // for this allocation is in the destructor for SqlTable.  clang-analyzer-cplusplus.NewDeleteLeaks identifies this
+  // as a potential leak and throws an error incorrectly.
+  // NOLINTNEXTLINE
+  tables_.Insert(schema.GetVersion(), {dt, layout, col_map, inv_col_map});
```

Comment:
Consider a user frequently updates the schema. Is there any chance for this map to shrink? Can some entries in the map be deleted when a table becomes empty and would never be used again?
Do we have to worry about this? Is GC responsible for this?

Reply:
At the moment there is no way to shrink the table (short of leveraging checkpoint/recovery to restart the process).  The stretch/longterm goal is to implement a compactor that would work on reducing the number of versions that exist.